### PR TITLE
Nan handling

### PR DESCRIPTION
**Pull request by rbtchun3**

### CODE REVIEW

File: src/ros_filter.cpp [hunk 1]
```diff
@@ -1916,6 +1916,15 @@ namespace RobotLocalization
...
               " This was likely due to poorly conditioned process, noise, or sensor covariances.");
+        if(last_odom.header.stamp.sec > 0.0)
+        {
+          ROS_ERROR_STREAM("Apply Previous Pose");
+          geometry_msgs::PoseWithCovarianceStamped last_pose;
+          last_pose.header = last_odom.header;
+          last_pose.pose = last_odom.pose;
+          setPoseCallback(boost::make_shared<geometry_msgs::PoseWithCovarianceStamped>(last_pose));
+          // we might want to early return
+        }
```
Comment (iandmil) [inline src/ros_filter.cpp:1928]: Maybe have an else condition that just sets state to 0?

File: src/ros_filter.cpp [hunk 2]
```diff
@@ -1916,6 +1916,15 @@ namespace RobotLocalization
...
         ROS_ERROR_STREAM("Critical Error, NaNs were detected in the output state of the filter." <<
               " This was likely due to poorly conditioned process, noise, or sensor covariances.");
+        if(last_odom.header.stamp.sec > 0.0)
+        {
+          ROS_ERROR_STREAM("Apply Previous Pose");
+          geometry_msgs::PoseWithCovarianceStamped last_pose;
+          last_pose.header = last_odom.header;
+          last_pose.pose = last_odom.pose;
+          setPoseCallback(boost::make_shared<geometry_msgs::PoseWithCovarianceStamped>(last_pose));
+          // we might want to early return
```
Comment (iandmil) [inline src/ros_filter.cpp:1926]: Hmm yeah, I think that's a good idea

**Review by iandmil — COMMENTED** [review]
Nice work, so much better than the old solution!

File: src/ros_filter.cpp [hunk 3]
```diff
@@ -1916,6 +1916,15 @@ namespace RobotLocalization
       {
         ROS_ERROR_STREAM("Critical Error, NaNs were detected in the output state of the filter." <<
               " This was likely due to poorly conditioned process, noise, or sensor covariances.");
+        if(last_odom.header.stamp.sec > 0.0)
+        {
+          ROS_ERROR_STREAM("Apply Previous Pose");
```
Comment (iandmil) [inline src/ros_filter.cpp]: Can we verify that this case gets triggered by the test?

File: src/ros_filter.cpp (hunk 1 above)
Comment (rbtchun3) [reply src/ros_filter.cpp:1928]: it's taken care of in the setposecallback

File: src/ros_filter.cpp (hunk 3 above)
Comment (rbtchun3) [reply src/ros_filter.cpp]: https://github.com/burro-robotics/burro/pull/4229#:~:text=with%20fix-,%24%20rostest%20burro_tests%20ukf_nan_handling.test%20%2D%2Dreuse%2Dmaster,-...%20logging%20to%20/home

File: src/ros_filter.cpp (hunk 2 above)
Comment (mona-robotics) [reply src/ros_filter.cpp:1926]: What's the upside/downside with early return versus publishing?

Comment (rbtchun3) [reply src/ros_filter.cpp:1926]: it likely doesn't matter as far as I read through the code.
but it is safer to do early return to fit the original workflow: detecting nan, reset the pose and resume in next update cycle

**Review by mona-robotics — APPROVED** [review]

File: src/ros_filter.cpp (hunk 2 above)
Comment (rbtchun3) [reply src/ros_filter.cpp:1926]: and both pass the integration tests

File: src/ros_filter.cpp (hunk 1 above)
Comment (iandmil) [reply src/ros_filter.cpp:1928]: How so?  Isn't setPoseCallback not even called if last_odom stamp is 0?

File: src/ros_filter.cpp (hunk 2 above)
Comment (SalimZRai) [reply src/ros_filter.cpp:1926]: it can be too much, but is there a way to unittest that in their repo?

File: src/ros_filter.cpp (hunk 1 above)
Comment (SalimZRai) [reply src/ros_filter.cpp:1928]: I remember we had bad experience using setpose service (lot of instabilities), do you think we can experience the same with the callback @rbtchun3 ?

Comment (rbtchun3) [reply src/ros_filter.cpp:1928]: oh @iandmil I missread. yeah that's doable but I haven't figured out how to run their unit test

Comment (rbtchun3) [reply src/ros_filter.cpp:1928]: @SalimZRai this way will not have the congestion issues as everything is running in linear sequence(I believe that's what causes instabilities of ROS as system struggles)

**Review by SalimZRai — COMMENTED** [review]
@rbtchun3 are you waiting for additional comments?

**Pull request merged**